ipalib/rpc.py

```diff
@@ -274,102 +274,171 @@ def xml_dumps(params, version, methodname=None, methodresponse=False,
+    """
+    result = _JSONPrimer(version).convert(val)
+    if pretty_print:
+        return json.dumps(result, indent=4, sort_keys=True)
     else:
-        return val
+        return json.dumps(result)
 
 
-def json_decode_binary(val):
-    '''
-    JSON cannot transport binary data. In order to transport binary data we
-    convert binary data to a form like this:
-
-   {'__base64__' : base64_encoding_of_binary_value}
-
-   see json_encode_binary()
-
-    After JSON had decoded the JSON stream back into a Python object we must
-    recursively scan the object looking for any dicts which might represent
-    binary values and replace the dict containing the base64 encoding of the
-    binary value with the decoded binary value. Unlike the encoding problem
-    where the input might consist of immutable object, all JSON decoded
-    container are mutable so the conversion could be done in place. However we
-    don't modify objects in place because of side effects which may be
-    dangerous. Thus we elect to spend a few more cycles and avoid the
-    possibility of unintended side effects in favor of robustness.
-    '''
+def _ipa_obj_hook(dct, _iteritems=six.iteritems, _list=list):
+    """JSON object hook
 
-    if isinstance(val, dict):
-        if '__base64__' in val:
-            return base64.b64decode(val['__base64__'])
-        elif '__datetime__' in val:
-            return datetime.datetime.strptime(val['__datetime__'],
-                                              LDAP_GENERALIZED_TIME_FORMAT)
-        elif '__dns_name__' in val:
-            return DNSName(val['__dns_name__'])
-        else:
-            return dict((k, json_decode_binary(v)) for k, v in val.items())
-    elif isinstance(val, list):
-        return tuple(json_decode_binary(v) for v in val)
+    :see: _JSONPrimer
+    """
+    if '__base64__' in dct:
+        return base64.b64decode(dct['__base64__'])
+    elif '__datetime__' in dct:
+        return datetime.datetime.strptime(dct['__datetime__'],
+                                          LDAP_GENERALIZED_TIME_FORMAT)
+    elif '__dns_name__' in dct:
+        return DNSName(dct['__dns_name__'])
     else:
-        if isinstance(val, bytes):
-            try:
-                return val.decode('utf-8')
-            except UnicodeDecodeError:
-                raise ConversionError(
-                    name=val,
-                    error='incorrect type'
-                )
-        else:
-            return val
+        # XXX tests assume tuples. Is this really necessary?
+        for k, v in _iteritems(dct):
+            if v.__class__ is _list:
+                dct[k] = tuple(v)
+        return dct
+
+
+def json_decode_binary(val):
+    """Convert serialized JSON string back to Python data structure
+
+    :param val: JSON string
+    :type val: str, bytes
+    :return: Python data structure
+    :see: _ipa_obj_hook, _JSONPrimer
+    """
+    if isinstance(val, bytes):
+        val = val.decode('utf-8')
```

Comment:
It should be handled by the JSON library if possible. The approach allows us to replace the JSON library with a more efficient one without breaking API.